src/sidebar/store/util.js

```diff
@@ -20,42 +20,62 @@ function actionTypes(updateFns) {
 function createReducer(...actionToUpdateFn) {
   // Combine the (action name => update function) maps together into a single
   // (action name => update functions) map.
+  //
+  // After namespace migration, remove the requirement for actionToUpdateFns
+  // to use arrays. Why? createReducer will be called once for each module rather
+  // than once for all modules.
   const actionToUpdateFns = {};
   actionToUpdateFn.forEach(map => {
     Object.keys(map).forEach(k => {
       actionToUpdateFns[k] = (actionToUpdateFns[k] || []).concat(map[k]);
     });
   });
 
-  return (state, action) => {
+  return (state = {}, action) => {
     const fns = actionToUpdateFns[action.type];
     if (!fns) {
       return state;
     }
+    // Some states return an array rather than an object. They need to be
+    // handled differently so we don't cast them to an object.
+    if (Array.isArray(state)) {
+      return [...fns[0](state, action)];
+    }
     return Object.assign({}, state, ...fns.map(f => f(state, action)));
   };
 }
 
 /**
- * Takes an object mapping keys to selector functions and the `getState()`
- * function from the store and returns an object with the same keys but where
- * the values are functions that call the original functions with the `state`
- * argument set to the current value of `getState()`
+ * Takes a mapping of namespaced modules and the store's `getState()` function
+ * and returns an aggregated flat object with all the selectors at the root
+ * level. The keys to this object are functions that call the original
+ * selectors with the `state` argument set to the current value of `getState()`
+ * for namespaced modules or `getState().base` for non-namespaced modules.
  */
-function bindSelectors(selectors, getState) {
-  return Object.keys(selectors).reduce(function(bound, key) {
-    const selector = selectors[key];
-    bound[key] = function() {
-      const args = [].slice.apply(arguments);
-      args.unshift(getState());
-      return selector.apply(null, args);
-    };
-    return bound;
-  }, {});
+function bindSelectors(namespaces, getState) {
+  const totalSelectors = {};
+  Object.keys(namespaces).forEach(namespace => {
+    const selectors = namespaces[namespace].selectors;
+    const scopeSelector = namespaces[namespace].scopeSelector;
+    Object.keys(selectors).forEach(selector => {
+      totalSelectors[selector] = function() {
+        const args = [].slice.apply(arguments);
+        if (scopeSelector) {
```

Comment:
when using scopeSelector, the selector does not need to drill into the top level namespace scope. This is a way to allow legacy modules to work without changes but keep them under the "base" namespace.

Only the legacy (non-namespaced) modules get `scopeSelector=true`. This will eventually be deprecated when all modules are switched over.

Reply:
We'll still need to come up with something for selectors that access state from multiple modules, or some equivalent for what those selectors do.

Reply:
namespaced selectors get the root store so thy will still have access to everything , but paths will change of course.